### PR TITLE
You Probably Don't Need Derived State

### PR DESCRIPTION
**Pull request by bvaughn**
Continuation of PR #931

[View formatted blog post](https://deploy-preview-934--reactjs.netlify.com/blog/2018/06/07/you-probably-dont-need-derived-state.html)

(All of the red CI marks below are because of a problem with Node+Sharp. They don't have anything to do with this particular PR.)

### TODO
- [ ] Add more Code Sandbox links after each section to demonstrate the points.

### CODE REVIEW

File: content/blog/2018-06-07-you-probably-dont-need-derived-state.md [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+
+The lifecycle `componentWillReceiveProps` has been around for a long time. Until recently, it was the only way to update state in response to a change in props without triggering an additional render. In version 16.3, [we introduced a replacement lifecycle, `getDerivedStateFromProps`](/blog/2018/03/29/react-v-16-3.html#component-lifecycle-changes), that served the same purpose. However, as we've helped people migrate their components to the new lifecycle, we've uncovered some common anti-patterns for derived state that result in subtle and confusing bugs.
+
+> Note
+>
+> The anti-patterns described in this post apply to both `componentWillReceiveProps` and `getDerivedStateFromProps`!
+
+In 16.4, we released a bugfix that [makes derived state behavior more predictable](https://github.com/facebook/react/issues/12898) so the results of misusing it are easier to notice. This blog post will guide you through when derived state should be used, and when there are better alternatives.
```
[the remark this answers is not in the document]
Comment (bvaughn) [reply content/blog/2018-06-07-you-probably-dont-need-derived-state.md]: I agree that it's similar to the opening summary.

I kind of like how it flows though, after the previous section. "We observed this problem, so we made this change to improve it". I worry that if we remove this section, the implied association between the previous paragraph and the intro blurb might not be obvious to everyone.

File: content/blog/2018-06-07-you-probably-dont-need-derived-state.md [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+
+> Note
+>
+> Even though the example above shows `componentWillReceiveProps`, the same anti-pattern applies to `getDerivedStateFromProps`.
+
+We've just made a big improvement. Now our component will no longer erase what we've typed when it is re-rendered. At this point, what we have is fairly similar to the [scroll direction example](blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props) we mentioned before.
+
+There is still one subtle problem though, and it's probably easiest to illustrate with an example. Let's say our `EmailInput` component is used inside of an "edit profile" form. The first time the form is rendered, the component will display our email address. Let's say we edited the form (including our email address) but then changed our mind and clicked a "reset" button. At this point, we would expect all form fields to return to their initial values— but the `EmailInput` component **will not be reset**. Do you know why?
+
+The reason for this is that `props.email` never actually changed in the above scenario. Both times the "edit profile" form rendered, it would pass the saved email address via props.
```
Comment (sw-yx) [inline content/blog/2018-06-07-you-probably-dont-need-derived-state.md]: I was with you until this para. Its not clear at all to me that `props.email` "never actually changed. You're just telling me that. I could blame it on poor implementation of the form's reset method. Can you make this clearer?

Comment (gaearon) [reply content/blog/2018-06-07-you-probably-dont-need-derived-state.md]: Yeah I think we’ll need codepens for all failing scenarios.

Comment (bvaughn) [reply content/blog/2018-06-07-you-probably-dont-need-derived-state.md]: I could probably make it clearer by adding the "committed" descriptor in there. Thanks.

Comment (bvaughn) [reply content/blog/2018-06-07-you-probably-dont-need-derived-state.md]: > The reason for this is that "committed" `props.email` never actually changed in the above scenario. (We never clicked "save".) So when the edit form re-rendered the input, it passed the same email address via props.

Comment (bvaughn) [reply content/blog/2018-06-07-you-probably-dont-need-derived-state.md]: Here's one I've created for this scenario: https://codesandbox.io/s/mz5w04xpz9

Comment (bvaughn) [reply content/blog/2018-06-07-you-probably-dont-need-derived-state.md]: And here's one for the Account Manager scenario: https://codesandbox.io/s/mz2lnkjkrx

**Comment by bvaughn** [comment]
Interesting. I actually don't think this blog post is very long (if you set aside the code examples, which I think are vital). I intentionally tried to keep it as short as possible, while still being thorough (explaining concepts that people I've spoken to have shown confusion about).

**Pull request merged**